### ISSUE TITLE
feature: write CRC checksums?

### ISSUE THREAD

**Issue by atomGit** [okdana/mp3crc#3]
sorry if i'm bombarding you with issues :) but the ability to write checksums would be a huge plus i think

my use case specifically is being able to pass a CRC test after trimming silence from the file with Mp3splt, however i'd guess that this feature could be helpful for others for various reasons

in my case i'm writing a [bash shell script](https://12bytes.org/projects/de-borking-mp3s-on-linux) which wraps several utilities used to perform various operations (convert, trim, fix header errors, integrity checking, etc.) and this would be really helpful

... unless you know of an existing program which can write the CRC without transcoding (i'm not aware of anything)

**Comment by okdana** [comment]
That does seem useful, and i'm not aware of any other tools that will do it either.

One thing that does concern me is that i remember having some trouble accurately detecting where the audio stream ends (since it may be followed by lyrics, APE tags, ID3v1 tags, &c., and i *think* the length value in the LAME header isn't always reliable...?). I'm sure that's just because i didn't know what i was doing, but i'd want to be extra sure that i do now before i start making destructive modifications to any files.

**Comment by atomGit** [comment]
a lot of CBR encoded mp3s don't even have a LAME/Xing header i don't think

does the CRC info not have a fixed length? and if so, can you not just overwrite it? if i understand correctly, you're already calc'ing a new checksum of the audio stream to compare with the existing CRC in the header

if that doesn't work, would decoding the file help? i think you can count on ffmpeg (and ffprobe) being installed on pretty much any mainstream Linux distro if that helps

forgive me if i'm being stupid because i know next to nothing about this - i'm just tossing out ideas

**Comment by okdana** [comment]
>a lot of CBR encoded mp3s don't even have a LAME/Xing header i don't think

Yeah, LAME does it (obv), but i don't think it's common with other encoders.

>does the CRC info not have a fixed length? and if so, can you not just overwrite it?

Yes. Updating the CRC itself shouldn't be too difficult, assuming the header already exists in the file. For files that don't already have a header, we'd have to make one up, and it'd have to be accurate to avoid breaking software that respects it, and that requires a whole bunch of extra information that `mp3sum` doesn't (and probably never will) have.

For the files that already have the header, we'd just need to be sure that the data used to calculate the music CRC is in fact the music, without any extra garbage at the end. I got it working pretty well on my own collection just by fixing bugs as i encountered them, but you're the first person i've ever found besides me who actually uses this, so who knows what issues it has with 'real-world' MP3s.

**Comment by atomGit** [comment]
> For files that don't already have a header, we'd have to ...

not implementing that is ok with me - i'm more interested in correcting the existing CRC after a split operation

**Comment by atomGit** [comment]
> i think the length value in the LAME header isn't always reliable...?

that question got me to poke around a bit and perhaps i may have found some helpful info....

[MPEG Audio Frame Header](https://www.codeproject.com/Articles/8295/MPEG-Audio-Frame-Header)

he seems to have found a way to accurately calc the number of frames both for CBR and VBR

he stated something that confused me...

> If the protection bit in the header is not set, the frame contains a 16 bit CRC

do you know what he's saying here? i'm translating this to mean that, if the bit is not set, then *every audio frame contains a CRC* - is that right? i ask because that doesn't seem to jive with what i'm seeing in software - mp3info, Diags, etc., usually do not report the protection bit being set, yet the LAME encoder does not add a CRC to the frames by default

a little while later that day...

from LAME on SF...

> -p
> 
> Turn on CRC error protection.
> It will add a cyclic redundancy check (CRC) code in each frame, allowing to detect transmission errors that could occur on the MP3 stream. However, it takes 16 bits per frame that would otherwise be used for encoding, and therefore will slightly reduce the sound quality.

**Comment by okdana** [comment]
On mobile right now, so i'm kind of going off memory/assumptions — not that i was ever an expert in the first place — but i think it's just confusing terminology.

For some reason, flipping the protection bit *off* (setting it to `0`) indicates that the frame is 'protected' (has a CRC), whilst flipping it *on* (setting it to `1`) indicates the opposite. I'm not familiar with the software output you mentioned, but i'm guessing that they're just displaying the state of that bit in terms that make sense to the user rather than literally reporting whether it's flipped on/`1` or off/`0`

The protection bit is in the frame header, so it only applies to the frame that header is a part of. AFAIK there's no reason you couldn't have one frame with a CRC followed by another without, though i'm not sure why that would ever be the case unless maybe (?) you had spliced two different files together. I'm assuming that when tools like that show you a single protection on/off value they're just looking at the first frame... but i'm not positive

**Comment by atomGit** [comment]
ok, that clears things up for me - i assumed that bit was in the lame/xing/"global" header rather than in every audio frame - i just wanted to make sure i was doing the right thing in my script when i suggest to discard an mp3 if the error protection bit is 'on' (because of potential degradation of quality)

that first part of my last post was for you btw - i thought it might be helpful in figuring out how to accurately find the last audio frame for your script ... just the fact that this is a problem at all i find odd - whoever wrote the mp3 spec is goofy imo
